BaragonAgentService/src/main/java/com/hubspot/baragon/agent/managers/AgentRequestManager.java

```diff
@@ -67,11 +71,42 @@ public AgentRequestManager(BaragonStateDatastore stateDatastore,
   }
 
   public List<AgentBatchResponseItem> processRequests(List<BaragonRequestBatchItem> batch) throws InterruptedException {
+    Map<String, Optional<BaragonRequest>> requests = new HashMap<>();
+
+    // Grab the existing upstreams at the start of this batch, and have apply() and revert() calls modify the list in-memory as we work through batch items
+    Map<String, Collection<UpstreamInfo>> existingUpstreamsForThisBatch = batch.stream()
+        .map(requestItem -> {
+          final Optional<BaragonRequest> maybeRequest = requestDatastore.getRequest(requestItem.getRequestId());
+
+          requests.put(requestItem.getRequestId(), maybeRequest);
+
+          Optional<BaragonService> oldService;
+
+          if (maybeRequest.isPresent()) {
+            oldService = getOldService(maybeRequest.get());
```

Comment:
should we also save the old service instead of having to look it up again in processRequest?